src/dimmwitted.cc

```diff
@@ -276,15 +279,18 @@ void DimmWitted::learn() {
   Timer t_total, t;
 
   double current_stepsize = opts.stepsize;
-  const std::unique_ptr<double[]> prev_weights(new double[nweight]);
-  const std::unique_ptr<float[]> delta(
-      is_distributed() ? new float[nweight] : nullptr);  // 32-bit float
-  COPY_ARRAY(infrs.weight_values.get(), nweight, prev_weights.get());
+  const std::unique_ptr<double[]> prev_weights(
+      !is_distributed ? new double[nweight] : nullptr);
+
+  if (!is_distributed) {
+    COPY_ARRAY(infrs.weight_values.get(), nweight, prev_weights.get());
```

Comment:
In a similar spirit, we could have a `COPY_ARRAY_IF_POSSIBLE` that gracefully handles `nullptr` on either src/dst side, and suppress these conditional branches?  Just trying to keep this part fairly high-level and push down the distributed vs. single mode.